File: backend/app/schemas/legal_documents.py
"""Схемы юридических текстов (BIZ-52 срез-5, Доп. №1 разд. 52.2)."""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.domains.reseller.legal import LegalDocumentKind
from app.schemas.base import BaseSchema


class LegalDocumentRead(BaseSchema):
    """Действующая редакция — то, что показывают человеку."""

    kind: LegalDocumentKind
    title: str
    body: str
    version: int
    #: Чей текст: `self` / `reseller`. Партнёру нужно видеть, что у клиентов
    #: действует именно его редакция, а не чужая.
    source: str
    published_at: datetime


class LegalDocumentSummary(BaseSchema):
    """Строка списка: что вообще есть у этого арендатора, без тела текста."""

    kind: LegalDocumentKind
    title: str
    version: int
    source: str
    published_at: datetime


class LegalDocumentList(BaseSchema):
    items: list[LegalDocumentSummary]


class LegalAcceptanceStatus(BaseSchema):
    """Состояние одного вида текста для текущего пользователя (срез-11)."""

    kind: LegalDocumentKind
    title: str | None = None
    #: Действующая редакция. `None` — текста нет вовсе, принимать нечего.
    current_version: int | None = None
    #: Принятая этим человеком редакция. `None` — не принимал.
    accepted_version: int | None = None
    accepted_at: datetime | None = None
    #: Принято ли ИМЕННО действующее.
    accepted: bool = False
    #: Принята прежняя редакция, вышла новая. Отдельный признак, а не «не
    #: принято»: человеку показывают «условия изменились», а не «примите
    #: оферту» — это разные сообщения.
    outdated: bool = False


class LegalAcceptanceState(BaseSchema):
    """Что требует подписи от текущего пользователя."""

    items: list[LegalAcceptanceStatus]
    #: Виды, которые ещё не приняты. Дублирует признаки строк намеренно: без
    #: этого поля интерфейсу пришлось бы вычислять «надо ли показывать баннер»
    #: самому, то есть завести вторую правду о готовности.
    pending: list[LegalDocumentKind]


class LegalAcceptanceRead(BaseSchema):
    """Записанное принятие."""

    kind: LegalDocumentKind
    doc_version: int
    source: str
    accepted_at: datetime


class LegalDocumentPublish(BaseSchema):
    """Публикация новой редакции.

    Номер версии НЕ принимается от клиента: его выдаёт сервер как «предыдущая
    + 1». Прими номер снаружи — и две публикации подряд смогут получить один
    номер, а «какая редакция действовала» перестанет иметь однозначный ответ.
    """

    title: str = Field(min_length=1, max_length=255)
    body: str = Field(min_length=1)

File: frontend/src/api/legalDocuments.ts
import { apiClient } from "@/api/client";

/** Юридические тексты арендатора (ТЗ Доп. №1 разд. 52.2, четвёртый пункт). */
export type LegalDocumentKind = "offer" | "privacy" | "consent";

export interface LegalDocumentSummary {
  kind: LegalDocumentKind;
  title: string;
  version: number;
  /** Чей текст: `self` | `reseller`. */
  source: string;
  published_at: string;
}

export interface LegalDocument extends LegalDocumentSummary {
  body: string;
}

/**
 * Список действующих текстов. БЕЗ токена: оферту и политику ПДн человек обязан
 * прочитать ДО входа, иначе ссылка на них бессмысленна.
 */
export const listLegalDocuments = async (): Promise<LegalDocumentSummary[]> =>
  (await apiClient.get<{ items: LegalDocumentSummary[] }>("/public/legal")).data
    .items;

export const getLegalDocument = async (
  kind: LegalDocumentKind,
): Promise<LegalDocument> =>
  (await apiClient.get<LegalDocument>(`/public/legal/${kind}`)).data;

/** Состояние принятия одного вида текста текущим пользователем (срез-11). */
export interface LegalAcceptanceStatus {
  kind: LegalDocumentKind;
  title: string | null;
  current_version: number | null;
  accepted_version: number | null;
  accepted_at: string | null;
  accepted: boolean;
  /** Принята прежняя редакция, вышла новая — «условия изменились». */
  outdated: boolean;
}

export interface LegalAcceptanceState {
  items: LegalAcceptanceStatus[];
  /** Что ещё требует подписи. Считает сервер: вычислять на фронте значит
   *  завести вторую правду о том, подписан ли документ. */
  pending: LegalDocumentKind[];
}

/** Что ждёт подписи. С токеном: подписывает конкретный человек. */
export const getLegalAcceptanceState = async (): Promise<LegalAcceptanceState> =>
  (await apiClient.get<LegalAcceptanceState>("/legal/acceptance")).data;

/** Принять действующую редакцию. Номер и текст сервер берёт сам. */
export const acceptLegalDocument = async (
  kind: LegalDocumentKind,
): Promise<void> => {
  await apiClient.post(`/legal/acceptance/${kind}`);
};
